File: w3/w3java/Gun/Gun.java
public class Gun {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;
    
    public Gun(String model, int capacity) {
        this.amount = 0;
        this.capacity = capacity;
        this.isReady = false;
        this.model = model;
        this.totalShots = 0;
    }
    
    public Gun() {
        this.amount = 0;
        this.capacity = 8;
        this.isReady = false;
        this.model = "Beretta";
        this.totalShots = 0;
    }
    
    public int getAmount() {
        return this.amount;
    }
    
    public int getCapacity() {
        return this.capacity;
    }
    
    public boolean ready() {
        return this.isReady;
    }
    
    public String getModel() {
        return this.model;
    }
    
    public int getTotalShots() {
        return this.totalShots;
    }
    
    public void prepare() {
        this.isReady = !this.isReady;
    }
    
    public void reload() {
        this.amount = this.capacity;
    }
    
    public void shoot() throws OutOfRoundsException, NotReadyException {
        if ( !ready() ) {
            throw new NotReadyException();
        }
        if ( this.amount == 0 ) {
            isReady = false;
            throw new OutOfRoundsException();
        }
        
        this.amount -= 1;
        this.totalShots += 1;
        System.out.println("Bang!");
    }
    
    public String toString() {
        return this.model + " " + "[Ready: " + this.isReady + "]" + " " + "[Amount: " + this.amount + "/" + this.capacity + "]" + " " + "[Total shots: " + this.totalShots + "]";
    }
}

File: w3/w3cpp/Gun/Gun.cpp
#include <iostream>
#include "Gun.h"

Gun::Gun(const std::string& model, int capacity) {
    this->amount = 0;
    this->capacity = capacity;
    this->isReady = false;
    this->model = model;
    this->totalShots = 0;
}

Gun::~Gun() {
    
}

int Gun::getAmount() const {
    return this->amount;
}

int Gun::getCapacity() const {
    return this->capacity;
}

bool Gun::ready() const {
    return this->isReady;
}

const std::string& Gun::getModel() const {
    return this->model;
}

int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    isReady = !isReady;
}

void Gun::reload() {
    amount = capacity;
}

void Gun::shoot() {
    if ( !ready() ) {
        throw NotReady();
    }
    
    if ( amount == 0 ) {
        isReady = false;
        throw OutOfRounds();
    }
    
    amount -= 1;
    totalShots += 1;
    std::cout << "Bang!" << std::endl;
}

std::ostream& operator<<(std::ostream& out, const Gun& gun) {
    out << gun.getModel() << ' ' << "[Amount: " << gun.getAmount() << '/' << gun.getCapacity() << ']' << ' ';
    if ( gun.ready() ) {
        out << "[Ready]";
    } else {
        out << "[Not ready]";
    }
    out << ' ' << "[Total shots: " << gun.getTotalShots() << ']';
    return out;
}